nestjs/src/courses/task-verifications/self-education.service.ts

```diff
@@ -63,19 +92,33 @@ export class SelfEducationService {
       },
     } = courseTask.task.attributes as SelfEducationAttributes;
 
-    const { id: courseTaskId, type: courseTaskType, maxScore } = courseTask;
+    if (studentAnswers.length !== numberOfQuestions) {
+      throw new BadRequestException(
+        `Number of submitted answers (${studentAnswers.length}) does not match the number of questions (${numberOfQuestions})`,
+      );
+    }
 
-    const verifications = await this.taskVerificationsRepository.find({
-      where: { studentId, courseTaskId },
-      order: { createdDate: 'DESC' },
-    });
+    // Check if all answer values are integers
+    if (studentAnswers.flatMap(a => (Array.isArray(a.value) ? a.value : [a.value])).some(a => !Number.isInteger(a))) {
+      throw new BadRequestException('Invalid answer value');
+    }
+
+    const submittedIndices = new Set(studentAnswers.map(a => a.index));
+    if (submittedIndices.size !== numberOfQuestions) {
+      throw new BadRequestException('Submitted answer indices must be unique');
+    }
+
+    for (let i = 0; i < numberOfQuestions; i++) {
+      if (!submittedIndices.has(i)) {
+        throw new BadRequestException(`Missing answer for question index ${i}`);
+      }
+    }
```

Comment:
```suggestion
    if ([...submittedIndices].some(i => i < 0 || i > answers.length - 1) {
      throw new BadRequestException(`The answer index ${i} is out of range`);
    }
```
Because we don't know exactly which questions from all questions FE showed to a student. Answers here are from task attributes - `courseTask.task.attributes` and they are all answers of all possible questions.

Reply:
The function below could be changed as well by checking for undefined to make double validation.
```
    const rightAnswersCount = studentAnswers
      .map(({ index, value }) => {
        const rightAnswer = this.sortAnswers(answers[index]);
        const userAnswer = this.sortAnswers(value);
        return rightAnswer !== undefined && userAnswer !== undefined && Number(rightAnswer === userAnswer);
      })
      .reduce((sum, value) => sum + value, 0);
```